spec/index.bs

```diff
@@ -2976,17 +2958,67 @@ type and each [=binding type=] has an associated [=internal usage=], given by th
+
+    <dl dfn-type=dfn dfn-for="binding cost">
+        : <dfn>uniformBuffers</dfn>
+        : <dfn>dynamicUniformBuffers</dfn>
+        : <dfn>storageBuffers</dfn>
+        : <dfn>dynamicStorageBuffers</dfn>
+        : <dfn>samplers</dfn>
+        : <dfn>sampledTextures</dfn>
+        : <dfn>storageTextures</dfn>
+    </dl>
+
+    The [=binding cost=] |cost| computed as follows:
+
+    1. Let |cost| be a [=binding cost=] of all zeros.
+    1. For each |entry| in |entries|:
+        1. If |entry|.{{GPUBindGroupLayoutEntry/buffer}}?.{{GPUBufferBindingLayout/type}}
+            is {{GPUBufferBindingType/"uniform"}}:
+            1. Add 1 to |cost|.[=binding cost/uniformBuffers=].
+            1. If |entry|.{{GPUBindGroupLayoutEntry/buffer}}?.{{GPUBufferBindingLayout/hasDynamicOffset}}
+                is `true`, add 1 to |cost|.[=binding cost/dynamicUniformBuffers=].
+        1. If |entry|.{{GPUBindGroupLayoutEntry/buffer}}?.{{GPUBufferBindingLayout/type}}
+            is {{GPUBufferBindingType/"storage"}}
+            or {{GPUBufferBindingType/"read-only-storage"}}:
+            1. Add 1 to |cost|.[=binding cost/storageBuffers=].
+            1. If |entry|.{{GPUBindGroupLayoutEntry/buffer}}?.{{GPUBufferBindingLayout/hasDynamicOffset}}
+                is `true`, add 1 to |cost|.[=binding cost/dynamicStorageBuffers=].
+        1. If |entry|.{{GPUBindGroupLayoutEntry/sampler}} is not `undefined`,
+            add 1 to |cost|.[=binding cost/samplers=].
+        1. If |entry|.{{GPUBindGroupLayoutEntry/texture}} is not `undefined`,
+            add 1 to |cost|.[=binding cost/sampledTextures=].
+        1. If |entry|.{{GPUBindGroupLayoutEntry/storageTexture}} is not `undefined`,
+            add 1 to |cost|.[=binding cost/storageTextures=].
+
+    Issue: Add cost of external textures.
+</div>
+
 <div algorithm>
-    To get the <dfn abstract-op>layout entry binding type</dfn> in a given {{GPUBindGroupLayoutEntry}} |entry|:
-
-    1. If |entry|.{{GPUBindGroupLayoutEntry/buffer}} is not `undefined`:
-        1. Return |entry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/type}}.
-    1. If |entry|.{{GPUBindGroupLayoutEntry/sampler}} is not `undefined`:
-        1. Return |entry|.{{GPUBindGroupLayoutEntry/sampler}}.{{GPUSamplerBindingLayout/type}}.
-    1. If |entry|.{{GPUBindGroupLayoutEntry/texture}} is not `undefined`:
-        1. Return |entry|.{{GPUBindGroupLayoutEntry/texture}}.{{GPUTextureBindingLayout/sampleType}}.
-    1. If |entry|.{{GPUBindGroupLayoutEntry/storageTexture}} is not `undefined`:
-        1. Return |entry|.{{GPUBindGroupLayoutEntry/storageTexture}}.{{GPUStorageTextureBindingLayout/access}}.
+    The [=list=] of {{GPUBindGroupLayoutEntry}} values |entries|
+    <dfn lt="exceed the binding cost limits">exceeds the binding cost limits</dfn>
+    of [=supported limits=] |limits|
+    if any of the following are true:
+
+    - Let |cost| be the [=binding cost=] of |entries|.
```

Comment:
It feels just a little weird to use the same cost structure for testing against both layout-level limits and stage-level limits. It's not really a problem, just surprised me at first.